overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func taskRecoverySystemSetup(t *state.Task) (*recoverySystemSetup, error) {
+	var setup recoverySystemSetup
+
+	err := t.Get("recovery-system-setup", &setup)
+	if err == nil {
+		return &setup, nil
+	}
+	if err != state.ErrNoState {
+		return nil, err
+	}
+	// find the task which holds the data
+	var id string
+	if err := t.Get("recovery-system-setup-task", &id); err != nil {
+		return nil, err
+	}
+	ts := t.State().Task(id)
+	if ts == nil {
+		return nil, fmt.Errorf("internal error: cannot find referenced task %v", id)
+	}
+	if err := ts.Get("recovery-system-setup", &setup); err != nil {
+		return nil, err
+	}
+	return &setup, nil
+}
+
+func setTaskRecoverySystemSetup(t *state.Task, setup *recoverySystemSetup) error {
+	if t.Has("recovery-system-setup") {
+		t.Set("recovery-system-setup", setup)
+		return nil
+	}
+	var id string
+	if err := t.Get("recovery-system-setup-task", &id); err != nil {
+		return err
+	}
+	ts := t.State().Task(id)
+	if ts == nil {
+		return fmt.Errorf("internal error: cannot find referenced task %v", id)
+	}
+	ts.Set("recovery-system-setup", setup)
+	return nil
+}
+
+func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+	isRemodel := remodelCtx.ForRemodeling()
+	groundDeviceCtx := remodelCtx.GroundContext()
+
+	model := groundDeviceCtx.Model()
+	if isRemodel {
+		model = remodelCtx.Model()
+	}
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
+	}
+	label := setup.Label
+
+	// get all infos
+	infoGetter := func(name string) (*snap.Info, bool, error) {
+		// snap may be present in the system in which case info comes
+		// from snapstate
+		info, err := snapstate.CurrentInfo(st, name)
+		if err == nil {
+			hash, _, err := asserts.SnapFileSHA3_384(info.MountFile())
+			if err != nil {
+				return nil, true, fmt.Errorf("cannot compute SHA3 of snap file: %v", err)
+			}
+			info.Sha3_384 = hash
+			return info, true, nil
+		}
+		if _, ok := err.(*snap.NotInstalledError); !ok {
+			return nil, false, err
+		}
+		logger.Debugf("requested info for not yet installed snap %q", name)
+		// TODO: handle remodel case in which snap may not be installed
+		// yet, and thus we need to pull info from snapsup of relevant
+		// download tasks
+		return nil, false, fmt.Errorf("not implemented")
+	}
+
+	db := assertstate.DB(st)
+	// 1. prepare recovery system from remodel snaps (or current snaps)
+	newFiles, systemDir, err := createSystemForModelFromValidatedSnaps(model, label, db, infoGetter)
```

Comment:
won't the new files logic do the wrong thing if we reboot in the middle of this and redo it?

Reply:
I have #10220 open which makes this fail early when the directory exists. I think we need to discuss what would be the best course of action here. If a reboot occurs while the recovery system is being assembled the whole thing will get scrapped.